### ISSUE TITLE
Code in README "Advanced Example: Output to pipe" subject to deadlock

### ISSUE THREAD

**Issue by ronen** [enkessler/childprocess#48]
The README suggests the following code for 'Advanced Example: Output to pipe'

```
r, w = IO.pipe

proc = ChildProcess.build("echo", "foo")
proc.io.stdout = proc.io.stderr = w
proc.start
proc.wait

w.close
r.read #=> "test\n"
```

While this example works as is, somebody who follows the example, naively replacing `echo foo` with a command that generates more than one IO buffer's worth of data, will find that they get deadlock: the child process can't go on until the parent reads the buffer but the parent won't read until the child process finishes.  E.g. try something like this:

```
 proc = ChildProcess.build("ruby", "-e", 'puts "x"*256000')
```

(This isn't hypothetical.  I'm coming here after debugging a deadlock in a third-party gem; the gem author had used the example code verbatim, and my use case had larger data than the gem author had evidently tried.)

I'm no pipe wizard, but I think a more robust solution would be to do something like:

```
 BUFSIZ = 8192  # or anything smaller than system IO buffer size

 r, w = IO.pipe

proc = ChildProcess.build("ruby", "-e", 'puts "x"*256000')
proc.io.stdout = proc.io.stderr = w
proc.start
w.close

s = []
begin
   s << r.readpartial(BUFSIZ) while true
rescue EOFError
   # nothing to do
end

proc.wait
s.join #=> "xxxxxx..........."
```

(Although for this particular example, in practice the %x[] operator does what's wanted:

```
%x[ruby -e #{'puts "x"*256000'.shellescape}]
```

which is what I used in the patch I submitted to that gem.)

Anyway, bottom line, I guess I'm suggesting either making a more robust example, or adding a comment that the sample code only works for simple cases, and directing readers elsewhere for more sophisticated uses.

Cheers
